Moved Google Api interactions behind abstraction layers

Pikkuninja opened the pull request:
Created Map and PoiApi factories that return objects that provide interfaces for their respective Google services. These factories are then injected to MapController, so they are easily swappable if needed.
Moved MapController to it's own file.

Pikkuninja committed:
Moved Maps to their own controller and service.

Pikkuninja committed:
Merge pull request #1 from APT-project/master

Pull changes from base

Pikkuninja committed:
Merge branch 'master' of https://github.com/Pikkuninja/city_info

Pikkuninja committed:
Made services to factories, moved Google Places api calls to API class that can be fetched from factory

Pikkuninja committed:
Moved MapController to fullscreen div, it's not needed in tilemode

antti-TIEA4 commented:
Non-code related feedback:
It might be less stress free in the long term to create a separate branch (e.g. `git checkout -b api-abstraction`) for each PR. Working on the `master` branch works too but if someone else's commits get accepted before your commits or your PR gets rejected it can get quite messy to sort out your own `master` branch. It can be done ofc but with separate branches it's easier.
But if this is something you already know and this is your preferred way to do things, please disregard :)

Pikkuninja commented:
Yeah that probably would be good when other people start doing commits too. I'll tweak this PR a bit tonight and then look more into branching.

Pikkuninja committed:
Made Google Maps API to load lazily

Made map/poiapi factories return promises that when resolved return the actual objects

Pikkuninja commented:
And again a good reason to utilize branching better. I didn't even do the tweak that I was originally intending to do as it was so problematic, but ended up doing some major refactoring that should be in it's own PR. But atleast now the thing won't crash if a Map object is created before the async load of Google APIs is done.

Btw the original tweak I intended was to make PoiApi a service rather than a factory. Being a singleton would make sense for a API wrapper class but then the attributions / copyrights container that Google ToS demands becomes annoying to handle.
